clang/lib/Sema/SemaDecl.cpp

```diff
@@ -12562,6 +12562,14 @@ void Sema::ActOnUninitializedDecl(Decl *RealDecl) {
     if (getLangOpts().OpenCL &&
         Var->getType().getAddressSpace() == LangAS::opencl_local)
       return;
+    // In SYCL ESIMD device code non-constant file scope variables can't be
+    // initialized.
+    // TODO add proper diagnostics for both SYCL and OpenCL paths
```

Comment:
This seems like a good opportunity to just diagnose these.

Reply:
Are you suggesting to implement the check with this PR?
It will take time that I don't have at this point, hence TODO. This is a big feature with a number of gaps which will be addressed incrementally over time.

Reply:
I think this would have to happen to unblock this PR.  The feature is not complete/ready for merge until you've done your proper semantic analysis.

Reply:
I will try.
Looks like OpenCL right above is not complete/ready too.

Reply:
Update: this code is engaged only for implicit initialization, which shouldn't happen for private globals neither, so I just return skipping implicit initialization code creation (as OpenCL does). The check for explicit initialization is added elsewhere in SemaDecl.cpp, plus a test.